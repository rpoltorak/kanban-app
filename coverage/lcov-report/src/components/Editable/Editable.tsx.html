<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/Editable/Editable.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/Editable</a> Editable.tsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">58.7% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>27/46</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">75% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>15/20</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">42.86% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>6/14</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">64.86% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>24/37</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108</td><td class="line-coverage quiet"><span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from "react";
&nbsp;
import "./Editable.scss";
&nbsp;
interface EditableProps {
  value: string;
  placeholder?: string;
  autoFocus?: boolean;
  minLength?: number;
  maxLength?: number;
  onChange: (value: string) =&gt; void;
}
&nbsp;
interface EditableState {
  value: string;
  isEdited: boolean;
  isError: boolean;
}
&nbsp;
export default class Editable extends React.PureComponent&lt;EditableProps, EditableState&gt; {
  public static defaultProps = {
    value: "",
  };
&nbsp;
  public state = {
    value: this.props.value,
    isEdited: !this.props.value,
    isError: false,
  };
&nbsp;
  public ref = React.createRef&lt;HTMLDivElement&gt;();
&nbsp;
  public componentDidMount = () =&gt; {
    if (this.ref.current) {
      this.ref.current.innerHTML = this.state.value;
      this.checkErrors();
      <span class="missing-if-branch" title="else path not taken" >E</span>if (this.props.autoFocus) {
        this.ref.current.focus();
      }
    }
  }
&nbsp;
  public toggleEdit = <span class="fstat-no" title="function not covered" >() =&gt; <span class="cstat-no" title="statement not covered" ></span>this.setState(<span class="fstat-no" title="function not covered" >(</span>state: EditableState) =&gt; <span class="cstat-no" title="statement not covered" >({ isEdited: !state.isEdited }))</span>;</span>
&nbsp;
  public save = <span class="fstat-no" title="function not covered" >() =&gt; {</span>
<span class="cstat-no" title="statement not covered" >    this.toggleEdit();</span>
<span class="cstat-no" title="statement not covered" >    this.props.onChange(this.state.value);</span>
  }
&nbsp;
  public reset = <span class="fstat-no" title="function not covered" >() =&gt; <span class="cstat-no" title="statement not covered" ></span>this.setState(<span class="fstat-no" title="function not covered" >(</span>state: EditableState) =&gt; <span class="cstat-no" title="statement not covered" >({ value: this.props.value }))</span>;</span>
&nbsp;
  public handleBlur = <span class="fstat-no" title="function not covered" >async (</span>event: React.FocusEvent&lt;HTMLDivElement&gt;) =&gt; {
<span class="cstat-no" title="statement not covered" >    event.persist();</span>
&nbsp;
<span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" >    await this.checkErrors();</span></span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    if (!this.ref.current) {</span>
<span class="cstat-no" title="statement not covered" >      return;</span>
    }
&nbsp;
<span class="cstat-no" title="statement not covered" >    if (this.state.isError) {</span>
<span class="cstat-no" title="statement not covered" >      event.preventDefault();</span>
<span class="cstat-no" title="statement not covered" >      this.ref.current.focus();</span>
    } else {
<span class="cstat-no" title="statement not covered" >      this.props.onChange(this.ref.current.innerHTML);</span>
<span class="cstat-no" title="statement not covered" >      this.save();</span>
    }
  }
&nbsp;
  public checkErrors = () =&gt; {
    this.setState((state: EditableState) =&gt; ({
      isError: (!!this.props.minLength &amp;&amp; state.value.length &lt; this.props.minLength) ||
        (!!this.props.maxLength &amp;&amp; state.value.length &gt;= this.props.maxLength),
    }));
  }
&nbsp;
  public handleChange = <span class="fstat-no" title="function not covered" >(</span>event: React.FormEvent&lt;HTMLDivElement&gt;) =&gt; {
    const { <span class="cstat-no" title="statement not covered" >innerText </span>} = event.target as HTMLElement;
&nbsp;
<span class="cstat-no" title="statement not covered" >    this.setState(<span class="fstat-no" title="function not covered" >(</span>state: EditableState) =&gt; <span class="cstat-no" title="statement not covered" >({ value: innerText }),</span> this.checkErrors);</span>
  }
&nbsp;
  public renderCounter = () =&gt; {
    return this.props.maxLength &amp;&amp; this.state.isEdited &amp;&amp; (
      &lt;div className="editable__counter"&gt;
        {this.state.value.length}/{this.props.maxLength}
      &lt;/div&gt;
    );
  }
&nbsp;
  public render() {
    return (
      &lt;div className={this.state.isError ? "editable__wrapper error" : "editable__wrapper"}&gt;
        &lt;div
          contentEditable={true}
          className="editable__input"
          ref={this.ref}
          placeholder={this.props.placeholder}
          defaultValue={this.state.value}
          onInput={this.handleChange}
          onBlur={this.handleBlur}
        /&gt;
        {this.renderCounter()}
      &lt;/div&gt;
    );
  }
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Nov 12 2018 23:33:21 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
