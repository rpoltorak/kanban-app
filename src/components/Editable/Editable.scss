.editable {
  &__input {
    font-size: 14px;
    line-height: 1.5;
    min-height: 20px;
    max-width: 100%;
    resize: none;
    padding: 2px 5px;

    &:focus {
      outline: 0;
      border-bottom: 1px solid #000;
    }

    &:empty:before {
      content:attr(placeholder);
      color: #aaa;
    }
  }

  &__wrapper {
    margin-bottom: 5px;
  }

  &__counter {
    font-size: 12px;
    text-align: right;
  }
}

.error {
  color: red;

  .editable {
    &__input {
      border-color: red;
    }
  }
}


