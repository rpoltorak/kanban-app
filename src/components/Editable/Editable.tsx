import React from "react";

import "./Editable.scss";

interface EditableProps {
  value: string;
  placeholder?: string;
  autoFocus?: boolean;
  minLength?: number;
  maxLength?: number;
  onChange: (value: string) => void;
}

interface EditableState {
  value: string;
  isEdited: boolean;
  isError: boolean;
}

export default class Editable extends React.PureComponent<EditableProps, EditableState> {
  public static defaultProps = {
    value: "",
  };

  public state = {
    value: this.props.value,
    isEdited: !this.props.value,
    isError: false,
  };

  public ref = React.createRef<HTMLDivElement>();

  public componentDidMount = () => {
    if (this.ref.current) {
      this.ref.current.innerHTML = this.state.value;
      this.checkErrors();
      if (this.props.autoFocus) {
        this.ref.current.focus();
      }
    }
  }

  public toggleEdit = () => this.setState((state: EditableState) => ({ isEdited: !state.isEdited }));

  public save = () => {
    this.toggleEdit();
    this.props.onChange(this.state.value);
  }

  public reset = () => this.setState((state: EditableState) => ({ value: this.props.value }));

  public handleBlur = async (event: React.FocusEvent<HTMLDivElement>) => {
    event.persist();

    await this.checkErrors();

    if (!this.ref.current) {
      return;
    }

    if (this.state.isError) {
      event.preventDefault();
      this.ref.current.focus();
    } else {
      this.props.onChange(this.ref.current.innerHTML);
      this.save();
    }
  }

  public checkErrors = () => {
    this.setState((state: EditableState) => ({
      isError: (!!this.props.minLength && state.value.length < this.props.minLength) ||
        (!!this.props.maxLength && state.value.length >= this.props.maxLength),
    }));
  }

  public handleChange = (event: React.FormEvent<HTMLDivElement>) => {
    const { innerText } = event.target as HTMLElement;

    this.setState((state: EditableState) => ({ value: innerText }), this.checkErrors);
  }

  public renderCounter = () => {
    return this.props.maxLength && this.state.isEdited && (
      <div className="editable__counter">
        {this.state.value.length}/{this.props.maxLength}
      </div>
    );
  }

  public render() {
    return (
      <div className={this.state.isError ? "editable__wrapper error" : "editable__wrapper"}>
        <div
          contentEditable={true}
          className="editable__input"
          ref={this.ref}
          placeholder={this.props.placeholder}
          defaultValue={this.state.value}
          onInput={this.handleChange}
          onBlur={this.handleBlur}
        />
        {this.renderCounter()}
      </div>
    );
  }
}
