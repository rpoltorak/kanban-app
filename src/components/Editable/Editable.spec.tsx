import React from "react";
import { shallow, mount } from "enzyme";

import Editable from "./Editable";

const value = "example value";

describe("Editable Component", () => {
  test("to match snapshot", () => {
    const editable = shallow(
      <Editable
        value={value}
        placeholder="value"
        onChange={jest.fn()}
        minLength={1}
        maxLength={5}
        autoFocus
      />,
    );

    expect(editable).toMatchSnapshot();
  });

  test("should render correctly when passed props", () => {
    const editable = mount(
      <Editable
        value={value}
        placeholder="value"
        onChange={jest.fn()}
        minLength={1}
        maxLength={5}
        autoFocus
      />,
    );

    const editableInput = editable.find(".editable__input");

    expect(editableInput.exists()).toBe(true);
    expect(editableInput.text()).toEqual(value);
  });

  test("should display error when passed value doesn't meet the requirements", () => {
    const maxLength = 5;

    const editable = mount(
      <Editable
        value={value}
        placeholder="value"
        onChange={jest.fn()}
        minLength={1}
        maxLength={maxLength}
        autoFocus
      />,
    );

    expect(editable.find(".error").exists()).toBe(true);

    editable.setState({ value: "" });

    expect(editable.find(".error").exists()).toBe(true);
  });

  test("should display counter with proper values when changing value", () => {
    const maxLength = 5;
    const nextValue = "12";

    const editable = mount(
      <Editable
        placeholder="value"
        onChange={jest.fn()}
        minLength={1}
        maxLength={maxLength}
        autoFocus
      />,
    );

    editable.setState({ value: "12" });

    const editableCounter = editable.find(".editable__counter");

    expect(editableCounter.exists()).toBe(true);
    expect(editableCounter.text()).toBe(`${nextValue.length}/${maxLength}`);
  });
});
