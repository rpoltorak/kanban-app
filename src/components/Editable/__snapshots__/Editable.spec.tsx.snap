// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editable Component to match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Editable
    autoFocus={true}
    maxLength={5}
    minLength={1}
    onChange={[MockFunction]}
    placeholder="value"
    value="example value"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="editable__input"
          contentEditable={true}
          defaultValue="example value"
          onBlur={[Function]}
          onInput={[Function]}
          placeholder="value"
        />,
        false,
      ],
      "className": "editable__wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "editable__input",
          "contentEditable": true,
          "defaultValue": "example value",
          "onBlur": [Function],
          "onInput": [Function],
          "placeholder": "value",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="editable__input"
            contentEditable={true}
            defaultValue="example value"
            onBlur={[Function]}
            onInput={[Function]}
            placeholder="value"
          />,
          false,
        ],
        "className": "editable__wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "editable__input",
            "contentEditable": true,
            "defaultValue": "example value",
            "onBlur": [Function],
            "onInput": [Function],
            "placeholder": "value",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
